lib/github_changelog_generator/gitlab_fetcher.rb

```diff
@@ -0,0 +1,449 @@
+          end
+        end
+        threads.each(&:join)
+        threads = []
+      end
+      nil
+    end
+
+    # Fetch tag time from repo
+    #
+    # @param [Hash] tag GitHub data item about a Tag
+    #
+    # @return [Time] time of specified tag
+    def fetch_date_of_tag(tag)
+      Time.parse(tag["commit"]["committed_date"])
+    end
+
+    # Fetch and cache comparison between two github refs
+    #
+    # @param [String] older The older sha/tag/branch.
+    # @param [String] newer The newer sha/tag/branch.
+    # @return [Hash] Github api response for comparison.
+    def fetch_compare(older, newer)
+      unless @compares["#{older}...#{newer}"]
+        compare_data = check_response { @client.compare(@project_id, older, newer || "HEAD") }
+        compare_data = stringify_keys_deep(compare_data.to_hash)
+        compare_data["commits"].each do |commit|
+          commit["sha"] = commit["id"]
+        end
+        # TODO: do not know what the equivalent for gitlab is
+        if compare_data["compare_same_ref"] == true
+          raise StandardError, "Sha #{older} and sha #{newer} are not related; please file a github-changelog-generator issues and describe how to replicate this issue."
+        end
+        @compares["#{older}...#{newer}"] = stringify_keys_deep(compare_data.to_hash)
+      end
+      @compares["#{older}...#{newer}"]
+    end
+
+    # Fetch commit for specified event
+    #
+    # @param [String] commit_id the SHA of a commit to fetch
+    # @return [Hash]
+    def fetch_commit(commit_id)
+      found = commits.find do |commit|
+        commit["sha"] == commit_id
+      end
+      if found
+        stringify_keys_deep(found.to_hash)
+      else
+        # cache miss; don't add to @commits because unsure of order.
+        check_response do
+          commit = @client.commit(@project_id, commit_id)
+          commit = stringify_keys_deep(commit.to_hash)
+          commit["sha"] = commit["id"]
+          commit
+        end
+      end
+    end
+
+    # Fetch all commits
+    #
+    # @return [Array] Commits in a repo.
+    def commits
+      if @commits.empty?
+        @client.commits(@project_id).auto_paginate do |new_commit|
+          new_commit = stringify_keys_deep(new_commit.to_hash)
+          new_commit["sha"] = new_commit["id"]
+          @commits.push(new_commit)
+        end
+      end
+      @commits
+    end
+
+    # Return the oldest commit in a repo
+    #
+    # @return [Hash] Oldest commit in the gitlab git history.
+    def oldest_commit
+      commits.last
+    end
+
+    # @return [String] Default branch of the repo
+    def default_branch
+      @default_branch ||= @client.project(@project_id)[:default_branch]
+    end
+
+    # Fetch all SHAs occurring in or before a given tag and add them to
+    # "shas_in_tag"
+    #
+    # @param [Array] tags The array of tags.
+    # @return [Nil] No return; tags are updated in-place.
```

Comment:
You can learn more about the `void` special return annotation in YARD here: https://github.com/lsegal/yard/blob/master/docs/GettingStarted.md#declaring-types

Reply:
Thanks, learning Ruby step-by-step ;)